northstar/src/runtime/state.rs

```diff
@@ -36,61 +30,56 @@ use npk::{
 };
 use std::{
     collections::{HashMap, HashSet},
-    fmt,
     path::{Path, PathBuf},
     result,
     sync::Arc,
 };
 use tokio::{sync::oneshot, task, time};
 
 #[derive(Debug)]
-pub(super) struct State<'a> {
+pub(super) struct State<'a, L>
+where
+    L: Launcher,
+{
     config: &'a Config,
-    minijail: Minijail<'a>,
+    launcher: L,
     mount_control: Arc<MountControl>,
     events_tx: EventTx,
     repositories: HashMap<RepositoryId, Repository>,
-    containers: HashMap<Container, MountedContainer>,
+    containers: HashMap<Container, MountedContainer<L::Process>>,
     /// Internal test repository tempdir
     #[cfg(feature = "hello")]
     internal_repository: tempfile::TempDir,
 }
 
 #[derive(Debug)]
-pub enum BlockDevice {
+pub(super) enum BlockDevice {
     Loopback(PathBuf),
     Verity(PathBuf),
 }
 
 #[derive(Debug)]
-pub(super) struct MountedContainer {
+pub(super) struct MountedContainer<P> {
     pub(super) container: Container,
     pub(super) manifest: Manifest,
     pub(super) root: PathBuf,
     pub(super) device: BlockDevice,
-    pub(super) process: Option<ProcessContext>,
-}
-
-impl fmt::Display for MountedContainer {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}:{}", self.manifest.name, self.manifest.version)
-    }
+    pub(super) process: Option<ProcessContext<P>>,
 }
 
 #[derive(Debug)]
-pub(super) struct ProcessContext {
+pub(super) struct ProcessContext<P> {
+    process: P,
     started: time::Instant,
-    process: Process,
     cgroups: Option<super::cgroups::CGroups>,
 }
 
-impl ProcessContext {
+impl<P: Process> ProcessContext<P> {
```

Comment:
shouldn't the restriction for the type parameter done in the definition of the `ProcessContext` struct?

Reply:
The definition of `ProcessContext` doesn't need the restriction. This is why it's just in the `impl` block to ensure that `Process::stop` is implemented for `P`